### DISCUSSION TITLE
Settings Dialog Tabs

### DISCUSSION THREAD

**Discussion by oleg68** [GrandOrgue/grandorgue#934] [General]
While I was working on MIDI settings, I noted, that the setting dialog tab structure is far from optimal.

There are some another discussions about it: #686 and #924.

I invite users to discuss here, what Settings page structure is better.

Now, I found the following changes should be done
1. Rename `Default and Initial Settings` to `Defaults and Paths`
2. Move `Settings Store` and `Cache Store` from the `Options` tab to the `Defaults and Paths` tab
3. Move settings of audio groups from the `Audio Groups` to the `Audio Output` tab
4. Delete the `Audio Groups` tab
5. Move the `MIDI Devices` and the `Initial MIDI Configuration` tabs just after the `Reverb` tab before `Organs`
6. Move the `Organ Packages` to the `Organs` tab (joining the two lists to the single list is quite complex so it is a subject of a separate isue)

This topic is just about the reorganising the dialog structure, without changing any functionality. Any feature requests should be separated from this topic.

I especially ask @larspalo, @rousseldenis, @Jean-Gabriel1982 and @vpoguru to tell your opinions, because you have already told some ideas in other topics.

**Comment by oleg68** [comment]
@vpoguru

> * merging the "MIDI Devices" and "Initial MIDI Configuration" tabs into a new tab "MIDI"?

It is not possible because both the tabs have too many elements and they wouldn't fit into one tab together.

**Comment by larspalo** [comment]
With the constraints of not modifying functionality, I have the following opinions.

1. Changing tab naming to 'Defaults and Paths' is excellent
2. Moving the Paths from the options tab to the renamed Defaults and Paths is excellent
3. Merging Audio groups to Audio Output, might work, but I think one needs to restructure it with the Sound output below Sound ports and the Audio groups taking the left part (where Sound output is now). Though, I think we need to make sure that both the Audio groups and Mapping output keep their scrollable containers as they could conceivably be quite long in some users setups.
4. Yes, naturally, if we're merging them
5. Ok, but they serve completely different purposes, so I don't think they should be merged
6. Maybe... But we could as well wait with this merging and instead, when time comes, take a complete reworking of both the Organ packages and Organs tabs that will include better management possibilities of all their associated files (settings and cache), and an option to remove all missing entries, and orphaned files.

**Comment by Jean-Gabriel1982** [comment]
Hello,

Sorry for the delay.
I find all of these changes excellent.

For my part, I am always for an optimization of the elements, so that there is the least possible interaction for the user.

I would also merge the "Reverb" and "Temperaments" Tab in "Options"

**Reply by oleg68** [reply]
> I would also merge the "Reverb" and "Temperaments" Tab in "Options"

I think only similar things may be merged in the single tab. "Reverb" and "Temperaments" are totally different things, so I won't merge them.

**Reply by Jean-Gabriel1982** [reply]
Yes, but those tabs are Options for Organ configuration ...

In Option Tab, you have a part "Enhancement". I think "Reverb" might be in Option Tab.
Ok for "Temperaments" i agree with you. Why not merging it with Organs Tab ?

**Reply by oleg68** [reply]
@larspalo what do you think about placing Reverb to the Option tab?

Because we are going to move `Settings Store` and `Cache Store` from there, some place will be free.

**Reply by larspalo** [reply]
@oleg68 I don't think there will be enough room there in a good way. And in the longer run I think we perhaps should find another place for it entirely. That is to separate all global settings from everything we want to be sampleset (or organ) settings, which I think that Reverb could actually benefit from becoming eventually.

If anything at all should be moved to the Options tab, I think that the default values of Volume and Metronome could be candidates. Then the Default and Paths tab could be just Paths.

**Reply by Jean-Gabriel1982** [reply]
> I don't think there will be enough room there in a good way. And in the longer run I think we perhaps should find another place for it entirely. That is to separate all global settings from everything we want to be sampleset (or organ) settings, which I think that Reverb could actually benefit from becoming eventually.
> 
> If anything at all should be moved to the Options tab, I think that the default values of Volume and Metronome could be candidates. Then the Default and Paths tab could be just Paths.

But then in the "Organ" tab, why not group the options like "Reverb", "Temperaments", "Enhancements" (from Option tab)?

Thus, in the "Organ" tab, by selecting an instrument, in the "Properties" the user could choose his preferred options, which would be attached to each organ.

Thus, each organ would have its own parameters, defined by the user.
This would avoid modifying the settings each time the user wants to change the instrument, everything would be automatically saved ... (but maybe it's stupid what I'm saying, the user just needs to " save the parameters "in the" File "menu ...

If not for MIDI, why not merge "Midi devices" and "Initial Midi configuration" into a "Midi" tab and add in this tab just a link (in the same way as "properties") "Initial Midi configuration"?

**Reply by larspalo** [reply]
> But then in the "Organ" tab, why not group the options like "Reverb", "Temperaments", "Enhancements" (from Option tab)?
> 
> Thus, in the "Organ" tab, by selecting an instrument, in the "Properties" the user could choose his preferred options, which would be attached to each organ.

I think it's better to have an organ specific settings dialog to handle everything that's specific to a single organ (sampleset). In the general settings only the global management should be present. The Organ tab (and perhaps eventually the Organ packages) should be used to manage all registered organs in GO. In the future I hope that all relevant files belonging to the organs (settings, cache etc) could be managed there, but not the actual settings for the organs - that should be done in it's own dialog, I think.

One could argue whether it would be beneficial to have different audio settings also following with what organ/sampleset is loaded or not. The plus side would be for users with hardware supporting a wide range of settings that could then match the best against what the sampleset have to use resources most efficiently and with the least amount of resampling needed. On the other side you have users that have hardware that requires a specific setup to perform the best and rather wants to adapt all loaded organs/samplesets to that. I have no strong opinions on that yet...

> If not for MIDI, why not merge "Midi devices" and "Initial Midi configuration" into a "Midi" tab and add in this tab just a link (in the same way as "properties") "Initial Midi configuration"?

They serve different purposes entirely. One is like the Audio tab which handle the available devices on the hardware level, the other is what defaults should be set for any sampleset the first time it's loaded/opened/registered.

**Reply by Jean-Gabriel1982** [reply]
In this case, maybe it would be better to think carefully about the organization of this "settings" menu before modifying things?

And why not if the GrandOrgue interface evolves towards something more modern and functional, integrate this new presentation at the same time?

Or as a first step, make the changes proposed by @oleg68 and gradually improve the program.
It would be a good solution.

**Comment by oleg68** [comment]
Thank all the people for participating in the discussion.

So I see the new tab structure as the following:

- `Options`
    - Remove `Paths` group (moved to the `Paths` tab)
    - Add `Volume` (from the `Paths` tab) to the left bottom
    - Add `Metronome` to the right bottom
- `Paths` (rename from Defaults and Initial Settings)
    - Remove the left column with `Volume` and `Metronome` (move to the `Options` tab)
    - Remove the `Paths` static box (because the whole tab is called `Paths` and no more other settings exist here)
- `Audio` (renamed from `Audio Output`)
    - Add the `Audio Group` static box at the left top (move from the `Audio Groups` tab)
    - Move the `Sound output` checkbox at the right top above the `Sound ports`
    - Change the vertical resizing proportion of the upper and the lower parts from 1:2 to 1:1
- `MIDI Devices`
- `Initial Midi Configuration`
- `Organs`
    - Add the `Organ packages` at the bottom (from the separate tab) without any reworking
- `Reverb`
- `Temperaments`

**Reply by larspalo** [reply]
Very good!

**Comment by oleg68** [comment]
I've created #948 for implementing this.

**Comment by oleg68** [comment]
The changes of the first two pages:

![OptionsPage0](https://user-images.githubusercontent.com/11586506/149016357-dfa92514-3093-4796-a261-aef765c797eb.png)
![PathsPage0](https://user-images.githubusercontent.com/11586506/149016367-e270bfc5-92f0-4688-9c7f-d0bafd69107b.png)

Are they acceptable? Or should I change anything?

**Reply by larspalo** [reply]
I think it's great, except perhaps that on the Paths tab you really don't need to have "Organ settings path:" (as Paths is the tab name), just "Organ settings:" or even "Settings store:" like it was earlier is enough.

**Reply by larspalo** [reply]
Or perhaps add "directory:" for Settings store and Cache store so that they read:
`Settings store directory:` and `Cache store directory:` like the other.

**Reply by oleg68** [reply]
I agree, that using `directory` in all cases is better than using `path` and `directory` time to time.

**Reply by vpoguru** [reply]
You can probably omit both expressions if the tab label is "Paths". Isn't it enough to speak of "Organ settings", "Cache", "Sample sets", "Organ packages", "Settings import/export", "Audio recording", "MIDI recording", and "MIDI player"?

**Reply by Jean-Gabriel1982** [reply]
For me it is ok with @larspalo (without "path")

**Comment by oleg68** [comment]
But I'd like to ask you about the `Paths` tab. What is better, to have an one column layout like

- Label1:
- Field1
- Lable2:
- Field2
- ...
- LableN:
- FieldN

where each Label ends with `directory`

or a two column layout:

- Lable1: Field1
- Label2: Field2
- ...
- LabelN: FieldN

where there are neigher `path` nor `directory` in the lables.

Personally I'd prefer the second one.

**Reply by Jean-Gabriel1982** [reply]
In two colums it seems better, more clear

**Reply by larspalo** [reply]
@oleg68 I don't have any hard preferences on one column or two columns, both will work just fine as you suggest. But I do think that we might need to add some border/margins around the items (rows), especially if you go with a 2 col layout. (add sizer flags: | wxTOP | wxBOTTOM)

**Comment by oleg68** [comment]
@larspalo Done

The next attempt:

![PathsPage1](https://user-images.githubusercontent.com/11586506/149203898-b42dfad2-429b-4467-96bc-1734e091c2b7.png)

I'm going to change the default values for `Organ cache` and `Organ settings` from ``<UserHomeDir>/GrandOrgueCache`` and ``UserHomeDir>/GrandOrgueData`` to ``<UserHomeDir>/GrandOrgue/Cache`` and ``UserHomeDir>/GrandOrgue/Data``.

**Reply by larspalo** [reply]
@oleg68 That makes sense. Just check that the folders behave correctly when using multiple instances too. (each named instance should have their own GrandOrgue/Cache-InstanceName and GrandOrgue/Data-InstanceName folder)

**Comment by oleg68** [comment]
The new `Audio` tab
![SettingsAudio0](https://user-images.githubusercontent.com/11586506/149620561-ed7edd74-9020-4806-a948-6d2efd10d5ed.png)

**Reply by larspalo** [reply]
Looking good!

**Comment by oleg68** [comment]
The new composed `Organs` Tab
![SettingsOrgans0](https://user-images.githubusercontent.com/11586506/149622920-12770af2-36cd-4094-a4bd-799b5bec357f.png)

**Reply by larspalo** [reply]
Looking good too!

**Reply by paolo2504** [reply]
> The new composed `Organs` Tab
> ![SettingsOrgans0](https://user-images.githubusercontent.com/11586506/149622920-12770af2-36cd-4094-a4bd-799b5bec357f.png)

I like all changes you made. I suggest one more time to possibly show cache file names on some column (if I remember well, only organ Id was present and didn’t really understand its usage). 
Maybe also “path” column in organ tab is not useful..I mean: the organ path is common for all organs so I don’t see why show this info.

**Comment by oleg68** [comment]
I've almost finished. Now I'm updating the help.

**Comment by vpoguru** [comment]
If you increase the default window size by only a few pixels, "Initial MIDI Setup" will not be cut off (on macOS).

![Screenshot1](https://user-images.githubusercontent.com/13820940/149757707-33f5ac33-be63-487a-ad1f-b9c4f498396a.jpg)

![Screenshot2](https://user-images.githubusercontent.com/13820940/149757746-b9c4a6e0-ee56-4b9f-b5ae-95eb502a9a64.jpg)

**Comment by oleg68** [comment]
@vpoguru The default size is calculated automatically, so I can not increase it.

Seems there is a difference between the tabbed panel implemenations in wx on Linux/Windows and in wx on OSx.

When all tab labels do not fit the window size, only a part of them are displayed under Linux or Windows with a capability of paging.

Under OSx, all tabs are displayed, but some of them are wrapped. So the only way to prevent wrapping is to reduce the label length.

@larspalo, @vpoguru, @Jean-Gabriel1982 Earlier the tab was called ``Inirtial MIDI Configuration``. I shorted it to ``Initial MIDI Setup`` but it is still quite long. What shorter label may I put?

The idea of ``MIDI Setup`` does not reflect the overridable nature of it, so I don't find it as good.

``MIDI Defaults`` may be confused with ``MIDI Devices``, because these two lables are very similar.

May be ``Default MIDI``?

**Reply by larspalo** [reply]
@oleg68 I don't mind "Default MIDI" as a shortened form, though I honestly find the "problem" to really be a non issue as it's obviously related to wxwidgets and not GO per se. We already have so many layout issues relating to different screen resolutions (mainly vertical issues for smaller screens), that require user workarounds on a more complex level, that this simple "issue" - that only requires the user to resize the window width slightly - is almost ridiculous...

Somewhere down the line we'll inevitably have to live with certain implementations that are not exactly perfect on all platforms and kinds of hardware, but they still basically "work" and are "usable".

Another, slightly more complex solution to this "issue", would be to store the user preferences of the settings dialog size and restore it from that every time the dialog is opened and fall back to the default auto calculation only when stored settings are not found. Then the users with large screens can adapt the dialog to their liking and have it restored like that every time they open it.

**Reply by Jean-Gabriel1982** [reply]
@oleg68 and MIDI config ?

I agree a little with the message of @larspalo by making a digression.  
I think that in the end, maybe we should think about an overhaul of the GO interface for something more modern (see message from Mikelectric to which I subscribe) or more functional maybe for a version 4 or 5 of GO,  major release. This could solve or at least the user screen resolution problems..

**Reply by oleg68** [reply]
> @oleg68 and MIDI config ?

`Initial` is more important than `Config`. Because all these settings may be overriden at the Organ level.

**Reply by Jean-Gabriel1982** [reply]
So, I propose Initial MIDI, or First config, or My organ (as this tab is for configuring user material (keyboards, buttons, ...)

**Reply by oleg68** [reply]
@larspalo 

> Another, slightly more complex solution to this "issue", would be to store the user preferences of the settings dialog size and restore it from that every time the dialog is opened and fall back to the default auto calculation only when stored settings are not found. Then the users with large screens can adapt the dialog to their liking and have it restored like that every time they open it.

I agree that it would be a good long-term solution.

But may we rename the tab to ``Initial MIDI`` now as a short-term solution?

**Reply by larspalo** [reply]
> But may we rename the tab to `Initial MIDI` now as a short-term solution?

Yes, of course, I have no issues with that.

**Comment by Jean-Gabriel1982** [comment]
Nice release !!
How to translate "Path" tab in French ? All other tabs are translated but not "Path". It should be "Chemin" :-)

**Reply by oleg68** [reply]
You may submit a PR on https://github.com/GrandOrgue/grandorgue/blob/master/po/fr.po

Your should provide the translation for the plural form

```
msgid "Paths"
msgstr "Chemins"
```

I invite you to revise the all translation for the lines 2964-2986 because some labels are not more exist

**Reply by Jean-Gabriel1982** [reply]
> You may submit a PR on https://github.com/GrandOrgue/grandorgue/blob/master/po/fr.po
> 
How i can do ?
I click on "edit the file" ? This will create a fork ..

> Your should provide the translation for the plural form
> 
> ```
> msgid "Paths"
> msgstr "Chemins"
> ```
> 
> I invite you to revise the all translation for the lines 2964-2986 because some labels are not more exist

Ok I will do that :)

**Reply by oleg68** [reply]
> I click on "edit the file" ? This will create a fork ..

Yes. It is a right way:

1. Create a fork
2. Enable the Github Actions at the `Actions` page in the fork for doing automatic builds
3. Make some changes in your fork
4. Look at the ``Actions`` whether the build is success
5. You can download and test the built GrandOrgue from the Actions page
6. If everything looks all right than click on the `Contribute` button in your fork and create a pull request.

**Reply by Jean-Gabriel1982** [reply]
I think i made a mistake somewhere ... oups

**Reply by Jean-Gabriel1982** [reply]
@oleg68 sorry it is not regarding Tabs, but i saw in the GO>Menu>Settings, "Settings" is still in English but i didn't see in the .po file where to translate ...

**Reply by oleg68** [reply]
`Settings...` is not translated by design, because if the current language was wrong then the user would not be able to find `Setting->Options`  and to change the language.

**Reply by Jean-Gabriel1982** [reply]
Ah ok i understand ...
